e2etest/rewarding_test.go

```diff
@@ -69,3 +87,460 @@ func TestBlockReward(t *testing.T) {
+					return false, nil
+				}
+
+				//add expected block/epoch reward
+				for h := preExpectHigh + 1; h <= curHigh; h++ {
+					//Add block reward to current block producer
+					blk, err := chains[0].GetBlockByHeight(h)
+					require.NoError(t, err)
+					exptUnclaimed[getRewardAddStr[blk.ProducerAddress()]] =
+						big.NewInt(0).Add(exptUnclaimed[getRewardAddStr[blk.ProducerAddress()]], blockReward)
+
+					//update Epoch rewards
+					epochNum := h / blocksPerEpoch
+					if epochNum > preEpochNum {
+						require.Equal(t, epochNum, preEpochNum+1)
+						preEpochNum = epochNum
+
+						//Add normal epoch reward
+						for i := 0; i < numNodes; i++ {
+							rewardAddrStr := identityset.Address(i + numNodes).String()
+							expectAfterEpoch := big.NewInt(0).Add(exptUnclaimed[rewardAddrStr], epRwdShares[rewardAddrStr])
+							exptUnclaimed[rewardAddrStr] = expectAfterEpoch
+						}
+						//Add foundation bonus
+						foundationBonusLastEpoch, err := rps[0].FoundationBonusLastEpoch(context.Background(), wss[0])
+						require.NoError(t, err)
+						foundationBonus, err := rps[0].FoundationBonus(context.Background(), wss[0])
+						require.NoError(t, err)
+						if epochNum <= foundationBonusLastEpoch {
+							for i := 0; i < numNodes; i++ {
+								rewardAddrStr := identityset.Address(i + numNodes).String()
+								expectAfterEpochBonus := big.NewInt(0).Add(exptUnclaimed[rewardAddrStr], foundationBonus)
+								exptUnclaimed[rewardAddrStr] = expectAfterEpochBonus
+							}
+						}
+
+					}
+					preExpectHigh = h
+				}
+
+				//check pending Claim actions, if a claim is executed, then adjust the expectation accordingly
+				updateExpectationWithPendingClaimList(t, chains[0], exptUnclaimed, claimedAmount, pendingClaimActions)
+				endPendingActNum := len(pendingClaimActions)
+
+				curHighCheck := chains[0].TipHeight()
+				preHeight = curHighCheck
+
+				//If chain height or pending action changes, we need to take snapshot again.
+				return curHigh == curHighCheck && startPendingActNum == endPendingActNum, nil
+
+			})
+			require.NoError(t, err)
+
+			//Comparing the expected and real unclaimed balance
+			for i := 0; i < numNodes; i++ {
+				rewardAddrStr := identityset.Address(i + numNodes).String()
+				fmt.Println("Server ", i, " ", rewardAddrStr,
+					" unclaimed ", unClaimedBalances[rewardAddrStr].String(), " height ", preHeight)
+				fmt.Println("Server ", i, " ", rewardAddrStr,
+					"  expected ", exptUnclaimed[rewardAddrStr].String())
+
+				require.Equal(t, exptUnclaimed[rewardAddrStr].String(), unClaimedBalances[rewardAddrStr].String())
+			}
+
+			// Perform a random claim and record the amount
+			for i := 0; i < numNodes; i++ {
+
+				var amount *big.Int
+				rewardAddrStr := identityset.Address(i + numNodes).String()
+				rewardPriKey := identityset.PrivateKey(i + numNodes)
+				expectedSuccess := true
+
+				rand.Seed(time.Now().UnixNano())
+				switch r := rand.Intn(7); r {
+				case 0:
```

Comment:
what does each int stand for?

Reply:
int r is a selector for claim action test cases. These test cases are commented under each switch-case branch.
For example, case 0 is to test claim 0 amount, case 1 is to test claim all balance, case 4 is to test claim negative amount , etc

Reply:
can we at least comment in the code? or better have enum to make these explicit.

Reply:
I will update with explicit enum type.